components/common/src/templating/config.rs

```diff
@@ -588,34 +565,150 @@ fn is_toml_value_a_table(key: &str, table: &toml::value::Table) -> bool {
-fn set_permissions<T: AsRef<Path>>(path: T, pkg: &Pkg) -> hcore::error::Result<()> {
+fn set_permissions<T: AsRef<Path>>(path: T, user: &str, group: &str) -> hcore::error::Result<()> {
     use crate::hcore::os::users;
     use crate::hcore::util::posix_perm;
 
     if users::can_run_services_as_svc_user() {
-        posix_perm::set_owner(path.as_ref(), &pkg.svc_user, &pkg.svc_group)?;
+        posix_perm::set_owner(path.as_ref(), &user, &group)?;
     }
-    posix_perm::set_permissions(path.as_ref(), CONFIG_PERMISSIONS)
+
+    let permissions = if path.as_ref().is_dir() {
+        CONFIG_DIR_PERMISSIONS
+    } else {
+        CONFIG_PERMISSIONS
+    };
+    posix_perm::set_permissions(path.as_ref(), permissions)
 }
 
 #[cfg(windows)]
-fn set_permissions<T: AsRef<Path>>(path: T, _pkg: &Pkg) -> hcore::error::Result<()> {
+fn set_permissions<T: AsRef<Path>>(path: T, _user: &str, _group: &str) -> hcore::error::Result<()> {
     use crate::hcore::util::win_perm;
 
     win_perm::harden_path(path.as_ref())
 }
 
+/// Entry point to recursively walk the configuration directory and subdirectories to
+/// construct the list of template files
+///
+/// `dir` should be a directory that exists.
+fn load_templates<T>(dir: T) -> Result<TemplateRenderer>
+where
+    T: AsRef<Path>,
+{
+    let mut template = TemplateRenderer::new();
+    if dir.as_ref().is_dir() {
+        load_templates_recurse(&dir, &PathBuf::new(), &mut template)?;
+    } else {
+        debug!(
+            "Configuration directory is not a directory: {:?}",
+            dir.as_ref()
+        );
+    }
+    Ok(template)
+}
+
+/// Helper function for `load_templates` to recurse into the directory structure.
+///
+/// `dir` should be a directory
+fn load_templates_recurse<T>(
+    dir: T,
+    context: &Path,
+    mut template: &mut TemplateRenderer,
+) -> Result<()>
+where
+    T: AsRef<Path>,
+{
+    if let Ok(entries) = std::fs::read_dir(dir) {
+        for entry in entries {
+            if let Ok(entry) = entry {
+                // Skip any entries in the template directory which aren't files or directories
+                match entry.file_type() {
+                    Ok(file_type) => {
+                        if file_type.is_file() {
+                            // We're storing the pathname relative to the input config directory
+                            // as the identifier for the template
+                            let name = entry.file_name().to_string_lossy().into_owned();
+                            let relative_path = context.join(&name);
+
+                            // JW TODO: This error needs improvement. TemplateFileError is too generic.
+                            template
+                                .register_template_file(
+                                    &relative_path.to_string_lossy(),
+                                    &entry.path(),
+                                )
+                                .map_err(Error::TemplateFileError)?;
```

Comment:
If we were ever going a lot of template processing, it might behoove us to avoid allocating all these new strings:
```rust
                if let Some(relative_path) = relative_path.to_str() {
                    template
                        .register_template_file(relative_path, &entry.path())
                        .map_err(Error::TemplateFileError)?;
                } else {
                    // Not sure whether we'd want to throw an error or just log something; should be very rare
                }
```

Reply:
`to_string_lossy()` returns a `Cow<str>`.  I thought that in the normal case where `relative_path` is a UTF-8 string then there would be no allocation ?  Did I misunderstand how `Cow` works ?

Reply:
After digging through the stdlib sources, you're totally right. I was getting thrown off by the convention that functions with `to` tend to allocate, whereas `as` don't.

That said, I'm not really certain if the path not being valid UTF-8 would work correctly in our system. Can you think of a situation where this would occur? If so, it's worth a test to see that this implementation would handle it, otherwise, I think throwing an error (or even just panicking) would be appropriate.

That said, this is a pretty edge case, and I don't think we have a thoughtful or consistent approach to handling the UTF-8 conversions errors generally, so if you left this as is, I think that's understandable.

Reply:
Reading around, I'm think this is a non-issue.  All major linux distros default to UTF-8 and it should be a problem on Windows. The only case I can think of is that someone creates a package on a non UTF-8 locale system with config filenames encoded using something other than UTF-8.  Then the user of the package runs it on a system with UTF-8 encoding. We'd end up with templates filenames which are different than the original file names.  But the probability of this all seems vanishingly small and I'm not sure some other part of the supervisor would not have broken first.

I'm for leaving this as-is as it was the behavior before this change, so we're not breaking anyone who might have relied on that behavior.  Also this doesn't seem to be the place to enforce UTF-8 filenames first in our code base.  If we made that decision then we should enforce in studio first and then it'd make sense to do a panic here as well.